File: backend/app/api/v1/schemas/retaceo.py
"""Pydantic DTOs for retaceo landed-cost allocations."""

from __future__ import annotations

import uuid
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field

from app.api.v1.schemas.common import ORMOut
from app.domain.entities.retaceo import RetaceoStatus


class RetaceoCreate(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    purchase_id: uuid.UUID
    total_freight: Decimal = Field(ge=0, max_digits=18, decimal_places=6)
    total_expenses: Decimal = Field(ge=0, max_digits=18, decimal_places=6)
    total_dai: Decimal = Field(ge=0, max_digits=18, decimal_places=6)
    import_vat: Decimal = Field(ge=0, max_digits=18, decimal_places=6)
    notes: str | None = Field(default=None, max_length=4000)


class RetaceoUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    total_freight: Decimal = Field(ge=0, max_digits=18, decimal_places=6)
    total_expenses: Decimal = Field(ge=0, max_digits=18, decimal_places=6)
    total_dai: Decimal = Field(ge=0, max_digits=18, decimal_places=6)
    import_vat: Decimal = Field(ge=0, max_digits=18, decimal_places=6)
    notes: str | None = Field(default=None, max_length=4000)


class RetaceoDetailResponse(ORMOut):
    id: uuid.UUID
    retaceo_id: uuid.UUID
    purchase_detail_id: uuid.UUID
    product_id: int
    unit_id: int
    quantity: Decimal
    cost_fob: Decimal
    freight: Decimal
    expenses: Decimal
    dai: Decimal
    total_cost: Decimal
    unit_cost: Decimal
    created_at: datetime | None = None
    updated_at: datetime | None = None


class RetaceoResponse(ORMOut):
    id: uuid.UUID
    company_id: uuid.UUID
    code: str
    purchase_id: uuid.UUID
    branch_id: uuid.UUID
    created_by_id: uuid.UUID
    currency: str
    details: tuple[RetaceoDetailResponse, ...]
    total_fob: Decimal
    total_freight: Decimal
    total_expenses: Decimal
    total_dai: Decimal
    import_vat: Decimal
    total_cost: Decimal
    freight_percentage: Decimal
    expense_percentage: Decimal
    dai_percentage: Decimal
    status: RetaceoStatus
    notes: str | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None


__all__ = [
    "RetaceoCreate",
    "RetaceoDetailResponse",
    "RetaceoResponse",
    "RetaceoUpdate",
]

File: frontend/src/lib/types/retaceo.ts
export type RetaceoStatus = 'draft' | 'calculated' | 'verified' | 'closed' | 'cancelled';

export interface RetaceoDetail {
  id: string;
  retaceo_id: string;
  purchase_detail_id: string;
  product_id: number;
  unit_id: number;
  quantity: string;
  cost_fob: string;
  freight: string;
  expenses: string;
  dai: string;
  total_cost: string;
  unit_cost: string;
  created_at: string | null;
  updated_at: string | null;
}

export interface Retaceo {
  id: string;
  company_id: string;
  code: string;
  purchase_id: string;
  branch_id: string;
  created_by_id: string;
  currency: string;
  details: RetaceoDetail[];
  total_fob: string;
  total_freight: string;
  total_expenses: string;
  total_dai: string;
  import_vat: string;
  total_cost: string;
  freight_percentage: string;
  expense_percentage: string;
  dai_percentage: string;
  status: RetaceoStatus;
  notes: string | null;
  created_at: string | null;
  updated_at: string | null;
}

export interface RetaceoCreateInput {
  purchase_id: string;
  total_freight: string;
  total_expenses: string;
  total_dai: string;
  import_vat: string;
  notes: string | null;
}

export interface RetaceoUpdateInput {
  total_freight: string;
  total_expenses: string;
  total_dai: string;
  import_vat: string;
  notes: string | null;
}

export interface RetaceoListParams {
  status?: RetaceoStatus;
  purchase_id?: string;
  branch_id?: string;
  page?: number;
  size?: number;
}
